tools/SourceKit/tools/sourcekitd/lib/API/sourcekitdAPI-InProc.cpp

```diff
@@ -0,0 +1,969 @@
+      case CustomBufferKind::CodeCompletionResultsArray:
+        return SOURCEKITD_VARIANT_TYPE_ARRAY;
+    }
+    llvm::report_fatal_error("sourcekitd object did not resolve to a known type");
+  }
+
+  SKDCustomData *asCustomData() override {
+    return this;
+  }
+
+  CustomBufferKind getKind() {
+    return Kind;
+  }
+
+  const void *getDataPtr() {
+    return BufferPtr->getBuffer().data();
+  }
+
+private:
+  CustomBufferKind Kind;
+  std::unique_ptr<llvm::MemoryBuffer> BufferPtr;
+};
+
+class SKDError: public SKDObject {
+public:
+  SKDError(sourcekitd_error_t Kind, StringRef Description)
+  : Kind(Kind), Description(Description.str()) {}
+
+  SKDError(SKDError const&) = delete;
+  SKDError &operator=(SKDError const&) = delete;
+
+  sourcekitd_variant_type_t getVariantType() override {
+    llvm::report_fatal_error("sourcekitd error object is not a variant type");
+  }
+
+  SKDError *asError() override {
+    return this;
+  }
+
+  sourcekitd_error_t getKind() {
+    return Kind;
+  }
+
+  StringRef getDescription() {
+    return StringRef(Description);
+  }
+
+private:
+  sourcekitd_error_t Kind;
+  std::string Description;
+};
+
+template <typename ImplClass, typename RetTy = void>
+class SKDObjectVisitor {
+public:
+  typedef std::vector<std::pair<UIdent, sourcekitd_object_t>> DictMap;
+
+  static bool compKeys(const std::pair<UIdent, sourcekitd_object_t> &LHS,
+                       const std::pair<UIdent, sourcekitd_object_t> &RHS) {
+    return sourcekitd::compareDictKeys(LHS.first, RHS.first);
+  }
+
+  RetTy visit(sourcekitd_object_t Obj) {
+    auto Object = static_cast<SKDObject *>(Obj);
+
+    if (auto DictionaryObject = Object->asDictionary()) {
+      DictMap Dict;
+      DictionaryObject->apply([&](sourcekitd_uid_t Key, SKDObject *Element){
+        UIdent UID = UIdentFromSKDUID(Key);
+        Dict.push_back({ UID, Element });
+        return true;
+      });
+      std::sort(Dict.begin(), Dict.end(), compKeys);
+      return static_cast<ImplClass*>(this)->visitDictionary(Dict);
+    }
+    if (auto ArrayObject = Object->asArray()) {
+      std::vector<sourcekitd_object_t> Vec;
+      ArrayObject->apply([&](size_t Index, SKDObject *Element){
+        Vec.push_back(Element);
+        return true;
+      });
+      return static_cast<ImplClass*>(this)->visitArray(Vec);
+    }
+    if (auto SKDUID = Object->getUID()) {
+      UIdent UID = UIdentFromSKDUID(SKDUID);
+      return static_cast<ImplClass*>(this)->visitUID(UID.getName());
+    }
+    auto OptString = Object->getString();
+    if (OptString.hasValue()) {
+      return static_cast<ImplClass*>(this)->visitString(StringRef(OptString.getValue()));
+    }
+    auto OptInt = Object->getInt64();
+    if (OptInt.hasValue()) {
+      return static_cast<ImplClass*>(this)->visitInt64(OptInt.getValue());
+    }
+    llvm_unreachable("unknown sourcekitd_object_t");
+  }
+};
+
+class SKDObjectPrinter : public SKDObjectVisitor<SKDObjectPrinter> {
+  raw_ostream &OS;
```

Comment:
I had tried to do that initially, but it turned out to be a little more complex than I had hoped. The only common sourcekitd API for retrieving values from a `sourcekitd_object_t` is the `RequestDict` class, which doesn't have a rich enough interface for this task currently. In particular, it doesn't have a mechanism for retrieving the set of keys  in the dictionary, or iterating over them. There wouldn't really be a way of dealing with a non-dictionary `sourcekitd_object_t` being passed into `sourcekitd::printRequestObject`.

How about just moving `SKDObjectPrinter` into the common file, since that code is actually identical. I suppose that would require adjusting the structure so it isn't a subtype of `SKDObjectVisitor` but rather relies on composition.